samples/encoder/main.cpp

```diff
@@ -0,0 +1,71 @@
+/**
+ * This is an example designed to work with the charge controller,
+ * but please ask someone before flashing the charge controller!
+*/
+
+#include <EVT/dev/encoder.hpp>
+#include <EVT/io/UART.hpp>
+#include <EVT/io/pin.hpp>
+#include <EVT/manager.hpp>
+#include <EVT/utils/log.hpp>
+#include <EVT/utils/time.hpp>
+
+namespace IO = EVT::core::IO;
+namespace DEV = EVT::core::DEV;
+namespace time = EVT::core::time;
+namespace log = EVT::core::log;
+
+constexpr IO::Pin A_PIN = IO::Pin::PA_8;
+constexpr IO::Pin B_PIN = IO::Pin::PA_9;
+
+constexpr uint8_t deviceCount = 1;
+IO::GPIO* devices[deviceCount];
+
+//Necessary static wrappers for the encoder to handle the pin interrupts
+DEV::Encoder* encoderWrapper;
```

Comment:
I would still like you to get rid of the global stuff here. To do this, you're going to have to modify `GPIO.hpp`, `GPIOf3xx.hpp`, and `GPIOf3xx.cpp`. You're going to want to add a `void*` to the accepted interrupt handler, and you're going to want to store the private data as well. Line 105 of `GPIO.hpp` should become the following: 
`virtual void registerIRQ(TriggerEdge edge, void (*irqHandler)(GPIO* pin, void* priv), void* priv) = 0;`

In `GPIOf3xx.cpp`, you're then going to have to modify `registerIRQ` and HAL_GPIO_EXTI_Callback. You're probably also going to have to add an array after `INTERRUPT_GPIOS` (line 13) that looks something like this:
`void* INTERRUPT_DATA[16] = {nullptr};`

Let me know if this doesn't make sense.